File: skillengine-frontend/src/app/core/services/skill.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  CreateSkillRequest,
  SkillNodeDto,
  SkillStackResponse,
} from '../models/skill.model';

@Injectable({ providedIn: 'root' })
export class SkillService {
  private readonly http = inject(HttpClient);
  private readonly skillsUrl = `${environment.apiBaseUrl}/skills`;
  private readonly adminUrl = `${environment.apiBaseUrl}/admin`;

  /** GET /api/v1/skills/related?query=&maxHops= */
  findRelated(query: string, maxHops = 3): Observable<SkillStackResponse[]> {
    const params = new HttpParams()
      .set('query', query)
      .set('maxHops', maxHops);

    return this.http.get<SkillStackResponse[]>(
      `${this.skillsUrl}/related`,
      { params }
    );
  }

  /** POST /api/v1/admin — ROLE_ADMIN only */
  createSkill(request: CreateSkillRequest): Observable<SkillNodeDto> {
    return this.http.post<SkillNodeDto>(this.adminUrl, request);
  }

  /** GET /api/v1/skills/search?query=&limit= — lightweight autocomplete */
  searchSkills(query: string, limit = 8): Observable<SkillNodeDto[]> {
    const params = new HttpParams()
      .set('query', query)
      .set('limit', limit);

    return this.http.get<SkillNodeDto[]>(
      `${this.skillsUrl}/search`,
      { params }
    );
  }
}

File: skillengine-backend/src/main/java/com/wexa/skillengine/service/SkillService.java
package com.wexa.skillengine.service;

import com.wexa.skillengine.dto.request.CreateSkillRequest;
import com.wexa.skillengine.dto.request.SkillPathRequest;
import com.wexa.skillengine.dto.response.SkillNodeDto;
import com.wexa.skillengine.dto.response.SkillPathResponse;
import com.wexa.skillengine.dto.response.SkillStackResponse;

import java.util.List;

public interface SkillService {

    List<SkillPathResponse> findPaths(SkillPathRequest request);

    SkillNodeDto createSkill(CreateSkillRequest request);

    List<SkillStackResponse> findRelated(String query, int maxHops);

    List<SkillNodeDto> searchSkills(String query, int limit);
}